---
title: "DEGGs vignette"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{DEGGs_vignette}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

# Overview

The DEGGs package enables the exploration of differential gene-gene correlations 
across phenotipical groups.  
This vignette uses RNA-seq data from breast cancer collected from 
The Cancer Genome Atlas program (TCGA).  
For this example we are going to compare two different breast cancer subgroups: 
HER2-positive and luminal A.  

Let's start by loading the package and data
```{r}
library(DEGGs)
data("BRCA_metadata")
data("BRCA_normCounts")
```

Make sure clinical and counts data are aligned

```{r}
identical(colnames(BRCA_normCounts), rownames(BRCA_metadata))
```

We can then proceed generating specific networks associated with the two subgroups.   
We will specify the groups of interest using the `subgroups` parameter.
If not specified all breast cancer subtypes will be compared (when more than two
groups are compared, a one-way ANOVA is used in place of the robust linear model).  
The provided count matrix uses entrez IDs, therefore we will set `entrezIDs = TRUE` 
and `convert_to_gene_symbols = TRUE` to visualise the matching gene symbols in our 
output. 

```{r}
subnetworks_object <- generate_subnetworks(normalised_counts = BRCA_normCounts, 
                                           metadata = BRCA_metadata,
                                           subgroup_variable = "SUBTYPE",
                                           subgroups = c("BRCA_Her2",
                                                         "BRCA_LumA"),
                                           entrezIDs = TRUE,
                                           convert_to_gene_symbols = TRUE) 
```
  
  
  
We use the `print_simple_network` function to plot and inspect the gene-gene
interaction network associated with the HER2-positive group.  
Links in red have p value < 0.05.  
```{r, fig.width = 7.7, fig.height = 6}
print_force_network(DEGGs_object = subnetworks_object, subgroup = "BRCA_Her2")
```
   
   
   
Alternatively, we can use igraph style network plotting.  
Significant links will be in blue.
```{r, fig.width = 5, fig.height = 5}
print_simple_network(DEGGs_object = subnetworks_object, subgroup = "BRCA_Her2")
```
  
  
While inspecting the network we might be interested in looking at some specific 
gene-gene interaction and how it really differs between the two groups.
For example, we look at the NOTCH2-DTX4 gene pair and plot their interacting 
regression models via the `print_regressions` function.  

```{r, fig.width = 7.2, fig.height = 4}
print_regressions(gene_A = "NOTCH2", gene_B = "DTX4", 
                  subgroups = c("BRCA_Her2", "BRCA_LumA"),
                  DEGGs_object = subnetworks_object, 
                  subgroup_variable = "SUBTYPE",
                  legend_offset = -0.55)
```
  
  
  
  
Finally, the same plot can be dinamically generated by clicking to the link when
usng the `View_interactive_subnetwork` shiny function.   
```{r, eval = FALSE, fig.width = 8, fig.height = 8}
View_interactive_subnetwork(DEGGs_object = subnetworks_object,
                            subgroup = "BRCA_Her2",
                            subgroup_variable = "SUBTYPE")
```
