---
title: "DEGGs vignette"
output: rmarkdown::html_vignette
fontsize: 12pt
vignette: >
  %\VignetteIndexEntry{DEGGs vignette}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

<style type="text/css">
.main-container {
  max-width: 768px;
  margin-left: auto;
  margin-right: auto;
}
</style> 

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

The DEGGs package enables the exploration of differential gene-gene correlations 
across phenotipical groups.  
This vignette uses breast cancer RNA-seq data from 
The Cancer Genome Atlas program ([TCGA](https://portal.gdc.cancer.gov/projects/TCGA-BRCA)).  
For this example we are going to compare two different breast cancer subgroups: 
HER2-positive and luminal A. 

## Installing from Github

```{r, eval=FALSE}
devtools::install_github("elisabettasciacca/DEGGs")
```

# Overview

Let's start by loading the package and data
```{r}
library(DEGGs)
data("BRCA_metadata")
data("BRCA_normCounts")
```

Make sure clinical and counts data are aligned

```{r}
identical(colnames(BRCA_normCounts), rownames(BRCA_metadata))
```

We can then proceed generating specific networks associated with the two subgroups.   
We will specify the groups of interest using the `subgroups` parameter.
If not specified all breast cancer subtypes will be compared (when more than two
groups are compared, a one-way ANOVA is used in place of the robust linear model).  
The provided count matrix uses entrez IDs, therefore we will set `entrezIDs = TRUE` 
and `convert_to_gene_symbols = TRUE` to visualise the matching gene symbols in our 
output. 

```{r}
subnetworks_object <- generate_subnetworks(normalised_counts = BRCA_normCounts, 
                                           metadata = BRCA_metadata,
                                           subgroup_variable = "SUBTYPE",
                                           subgroups = c("BRCA_Her2",
                                                         "BRCA_LumA"),
                                           entrezIDs = TRUE,
                                           convert_to_gene_symbols = TRUE) 
```
  
# Visualisation
To inspect the network in a shiny-style interactive way, we use the 
`View_interactive_subnetwork` function.   
```{r, eval=FALSE}
View_interactive_subnetwork(subnetworks_object)
```
   
This will let you navigate the networks associated to each group (1), 
filter by edge significance (2), and search for specific genes (3).   
  
<p>&nbsp;</p>
![](Shiny_overview.png){width=60%}
 <p>&nbsp;</p>
   
When clicking on a link, the related gene-gene regression plot will appear on 
the left. 
 <p>&nbsp;</p>
![](DEGGs_node_boxplot.PNG){width=60%}
<p>&nbsp;</p>   
   
When clicking on a node, a box plot will visualise the expression of the single 
node across the subgroups. 
 <p>&nbsp;</p>
 ![](DEGGs_link_plot.PNG){width=50%}
  <p>&nbsp;</p>
  
  
  
## List significant gene-gene interactions  
To obtain a table of all the significant gene-gene pairs found across groups,
you will need to use the `extract_sig_deggs` with `subnetworks_object` as input: 
```{r}
head(extract_sig_deggs(subnetworks_object))
```

The listed genes can be used as features of machine learning models 
(see *Sciacca, Elisabetta, et al. "Harnessing gene-gene interactions via an RNA-Sequencing network analysis framework improves precision medicine prediction in rheumatoid arthritis." bioRxiv (2021)*).


## Single regression plots 
If you want to plot the regression fits for a single gene-gene interaction 
you can use the `print_regressions` function specifying the names of the two genes:
  
```{r, fig.width = 6.5, fig.height = 4.5}
print_regressions(gene_A = "NOTCH2", gene_B = "DTX4",
                  deggs_object = subnetworks_object,
                  legend_position = "bottomright")
```
  
  
